### ISSUE TITLE
NASNet Large achieves low accuracy on ImageNet subset

### ISSUE THREAD

**Issue by andraskuti** [keras-team/keras#9586]
(As the issue is related to making predictions on a subset of ImageNet pictures I can not include a code that does not require external data download)

I make class predictions on the following dataset, which is a dog only subset of ImageNet images:
https://www.kaggle.com/c/dog-breed-identification/data

**Image preprocessing**
- Image size: 331*331 pixels
- Image preprocessing: preprocess_input() from keras.applications.inception_v3 
- Data augmentation: No

**Models**
- I tested the below listed Keras models with “include top” option, and ImageNet weights, no modification whatsoever
- I use predict_generator() to make the predictions

**Top 1 accuracy results**
- NASNetLarge: !!!51.21%!!!
- InceptionResNetV2: 92.61%
- Xception: 92.31%

**Environment:**
- Ububtu 17
- Keras 2.1.4
- Cuda V8.0.61
- CuDNN 7.0.5
- Tensorflow 1.4.0rc0
- NVIDIA Quadro P5000

**Comment by ahundt** [comment]
@andraskuti can you try out the version in keras-contrib? It was implemented independently which could help figure out the problem https://github.com/keras-team/keras-contrib/blob/master/keras_contrib/applications/nasnet.py

**Comment by andraskuti** [comment]
Sure. I got very similar results.
Top1 accuracy on the given task for NASNetLarge (keras-contrib version) is **50.61%**
(using the same setup as described above)

**Comment by titu1994** [comment]
This is surprising. I know this is starting to sound repetitive, bit can you try to make predictions using the Tensorflow NASNet large model directly using the slim implement?

If that can score close to the inception models, and the Keras models can't, it may point to some preprocessing bug, or perhaps even some bug in how the weights are loaded. 

When I was loading the weights, I was careful to check the shape, and number of weights that I loaded. All weights extracted as numpy arrays we're loaded successfully into their Keras equivalent layers. I also made sure to use the correct batch norm hyperparameters for the different models. Theoretically, there should be no major difference to the Keras version and the Tensorflow slim version

**Comment by ahundt** [comment]
I did try nasnet on a completely different task and get surprisingly poor results, but that could have been due to one of many factors, so I didn't suspect a problem like you describe here. If this is an actual conversion bug I'm curious what the root cause might be.

**Comment by ahundt** [comment]
We have another report of problems with the weights transferred for NASNet:
https://github.com/keras-team/keras/issues/9674

**Comment by ahundt** [comment]
There are several independent reports of problems with the NASNet weights, see https://github.com/keras-team/keras/issues/9586

**Comment by titu1994** [comment]
Yes I get that this is an issue. Presently, I am occupied with my final semester so I won't have time to dedicate to this. 

If anyone wants to take a look at how to fix this, please go to : https://github.com/titu1994/Keras-NASNet/tree/master/weight_translation

Create two folders, checkpoint and weights. Place the checkpoint you want to translate into it's folder and run `extract_weights.py`. Then, simply run the NASNet.py code after modifications to get the model you want and save the weights. 

Then you can cut paste these weights and load them into the normal model at the root and see if the predictions are correct or not.

**Comment by titu1994** [comment]
Also, there are other independent implementations of NASNet in Keras and PyTorch. Are they using these wonky weights as well ? If not, perhaps we could try using those weights and try some tests to see if model training or ImageNet evaluation is still an issue

**Comment by veronikayurchuk** [comment]
Hi all! I fixed it for NASNet-Mobile.
@titu1994, please check https://github.com/titu1994/Keras-NASNet/pull/12
I think, the same would work for NASNet-Large. (hope, I will fix it later)

Without the fix, weights were assigned incorrectly by type. 
For example, weights from conv0_bn/moving_mean were assigned to conv0_weights or
 weights from conv0_bn/beta were assigned to conv0_bn/moving_mean, etc.

Hope, it would help!

**Comment by titu1994** [comment]
@veronikayurchuk Thank you for this fix ! I am astonished that I thought for a second glob.glob() provides a sorted list of files. 

However, I tested out your weights from your repository on the evaluation script on the cheetah and elephant images. Your weights are providing exceedingly low scores for the cheetah, and the elephant (at 38% and 72% respectively).

I used your updated scripts for weight loading and created the weights for NASNet mobile again, and tried prediction on the above images and here are the new results : 

```python
Loaded NASNet Mobile
1/1 [==============================] - 4s 4s/step
Predicted: [[('n02130308', 'cheetah', 0.9362247), ('n02128385', 'leopard', 0.002087888), ('n02128925', 'jaguar', 0.0011769015), ('n07802026', 'hay', 0.00055119535), ('n02127052', 'lynx', 0.00042987464)]]
```

```python
Loaded NASNet Mobile
1/1 [==============================] - 4s 4s/step
Predicted: [[('n02504458', 'African_elephant', 0.78309786), ('n01871265', 'tusker', 0.124640346), ('n02504013', 'Indian_elephant', 0.016551586), ('n02391049', 'zebra', 0.0006357454), ('n01768244', 'trilobite', 0.0005858325)]]
```

After recreating the weights with your fix locally, the cheetah recognition skyrockets to 93%, whereas elephant increases to 78% as well.

Can you run these two images locally and post your scores using your own weights? I don't believe there should be such a drastic difference, especially since I basically ported the weights using your corrected code.

Also, can I inquire as to the OS, TF and Keras versions you use locally?

**Comment by titu1994** [comment]
What's more surprising however, is that NASNet large tends to perform worse on the same images as compared to the Mobile versions, indicating this fix may not entirely solve this problem.

Most probably, there is some weights which are being skipped, or some other problem with how the weights are loaded into their respective layers. @veronikayurchuk I've pushed a few patches to my repo, could you refer to them and see if you can try to find a fix for NASNet Large?

```python
Loaded NASNet Large
1/1 [==============================] - 5s 5s/step
Predicted: [[('n02130308', 'cheetah', 0.869967), ('n02128925', 'jaguar', 0.016802093), ('n02128385', 'leopard', 0.012212922), ('n02129165', 'lion', 0.0035213656), ('n02128757', 'snow_leopard', 0.003336287)]]
```

```python
Loaded NASNet Large
1/1 [==============================] - 7s 7s/step
Predicted: [[('n02504458', 'African_elephant', 0.52621555), ('n01871265', 'tusker', 0.33077866), ('n02504013', 'Indian_elephant', 0.02492782), ('n02093647', 'Bedlington_terrier', 0.00040763052), ('n01704323', 'triceratops', 0.00040301887)]]
```

**Comment by veronikayurchuk** [comment]
Hi, @titu1994 !

It sounds very weird, as I used my weights and the result are pretty high, but differs from yours.

Cheetah prediction is
`1/1 [==============================] - 3s 3s/step
Predicted: [[('n02130308', 'cheetah', 0.93829703), ('n02128385', 'leopard', 0.0031160081), ('n02128925', 'jaguar', 0.0016547953), ('n02128757', 'snow_leopard', 0.00067107304), ('n02127052', 'lynx', 0.0005479519)]]
`

Elephant prediction is
`1/1 [==============================] - 3s 3s/step
Predicted: [[('n02504458', 'African_elephant', 0.73011684), ('n01871265', 'tusker', 0.17269203), ('n02504013', 'Indian_elephant', 0.017120298), ('n02391049', 'zebra', 0.0007252202), ('n01768244', 'trilobite', 0.0006577261)]]`


I am using python3 and Ubuntu 16, keras - '2.1.5', tensorflow -  '1.4.0'.
Can you, please, tell us what is you Keras, TF, OS version?
:)

**Comment by titu1994** [comment]
Windows 10, CUDA 9, Keras 2.1.5, Tensorflow 1.6

**Comment by titu1994** [comment]
Are these scores for NASNet Mobile or NASNet Large? Cause your scores for cheetah match the ones from my mobile model. My scores for the elephant are slightly higher.

**Comment by veronikayurchuk** [comment]
yes, exactly!
It is for NASNet Mobile.
But, it sounds weird to me, that the score for elephant differs for 5,3%

**Comment by titu1994** [comment]
Hmm, are you using `from weight_translation.weight_load_mobile import *` and not `from weight_translation.weight_load_large import *` by accident ? 

In hindsight, using two files with the same methods was a bad design choice

**Comment by ahundt** [comment]
@titu1994 have these changes been brought into keras-contrib?

**Comment by titu1994** [comment]
Not yet I'm afraid. Seems the predictions do not line up with what Tensorflow slim version gets. 

I think porting the weights from the tf-hub model / setting it as the baseline and tuning the model weights might be a way forward. 

Recent discussion on the MobileNetV2 merge in Keras master can possibly point a proper way to transfer weights to models without the manual error prome method. 

If someone can try that, I would be grateful.

**Comment by titu1994** [comment]
@ahundt I've been playing around with TF-Hub to see what exactly is the issue here, so that I can fix NASNet.

Here are the script that i'm using to compare the two (based on the MobileNetV2 discussion a day or two ago) : 
```python
import numpy as np
import tensorflow as tf

import tensorflow_hub as hub

from keras.preprocessing import image
from keras.applications.imagenet_utils import decode_predictions

from nasnet import NASNetLarge, NASNetMobile, preprocess_input

size = 331

img_path = 'images/elephant.jpg'
img = image.load_img(img_path, target_size=(size, size))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)

y = np.copy(x)
x /= 255.0

inputs = tf.placeholder(tf.float32, [None, size, size, 3])
model = hub.Module("https://tfhub.dev/google/imagenet/nasnet_large/classification/1")
features = model(inputs, signature="image_classification", as_dict=True)
probs = tf.nn.softmax(features['default'])

with tf.variable_scope('keras'):
    model2 = NASNetLarge(weights=None)
    model2.load_weights('weights/NASNet-large.h5')

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    preds = sess.run(probs, {inputs: x})

y = preprocess_input(y)
preds2 = model2.predict(y)

print('Hub Predicted:', decode_predictions(preds[:, 1:]))
print('Model Predicted:', decode_predictions(preds2))
```

Heres the scores for the cheetah and elephant in my repo : 
```
Cheetah 
Hub Predicted: [[('n02130308', 'cheetah', 0.914707), ('n02128925', 'jaguar', 0.0027237295), ('n02128385', 'leopard', 0.0025306053), ('n02423022', 'gazelle', 0.0006407299), ('n01582220', 'magpie', 0.0005725752)]]
Model Predicted: [[('n02130308', 'cheetah', 0.89602923), ('n02128925', 'jaguar', 0.011666419), ('n02128385', 'leopard', 0.00976224), ('n02129165', 'lion', 0.0029320156), ('n02128757', 'snow_leopard', 0.0021664116)]]

Elephant
Hub Predicted: [[('n02504458', 'African_elephant', 0.5226582), ('n01871265', 'tusker', 0.30211642), ('n02504013', 'Indian_elephant', 0.05298932), ('n02408429', 'water_buffalo', 0.0003982857), ('n02391049', 'zebra', 0.00037648008)]]
Model Predicted: [[('n02504458', 'African_elephant', 0.5382313), ('n01871265', 'tusker', 0.31587833), ('n02504013', 'Indian_elephant', 0.026553808), ('n02093647', 'Bedlington_terrier', 0.00042403332), ('n02397096', 'warthog', 0.00039423758)]]
```

I'm going to try to port the weights using the technique suggested in that thread, and see if these results are the same are not. For now, the weights are very slightly different for sure.

**Comment by ahundt** [comment]
Hmm, that's unfortunate. I'd say the lack of human usability is one of the biggest problems with these NN designed NNs! If only there were a `human_usability` loss function. 👍

**Comment by taehoonlee** [comment]
The error has been fixed in #10209.
